src/plugins/data/public/data_sources/datasource_selector/datasource_selectable.tsx

```diff
@@ -0,0 +1,131 @@
+  useEffect(() => {
+    // Fetches datasets for a given datasource and returns it along with the source.
+    const fetchDataSetWithSource = async (ds: GenericDataSource): Promise<DataSetWithSource> => {
+      const dataSet = await ds.getDataSet();
+      return {
+        ds,
+        data_sets: dataSet,
+      };
+    };
+
+    // Map through all data sources and fetch their respective datasets.
+    const fetchDataSets = () => dataSources.map((ds: DataSourceType) => fetchDataSetWithSource(ds));
+
+    const isIndexPatterns = (dataset: string | IndexPatternOption): boolean => {
+      if (typeof dataset === 'string') return false;
+
+      return dataset.title !== undefined && dataset.id !== undefined;
+    };
+
+    // Get the option format for the combo box from the dataSource and dataSet.
+    const getSourceOptions = (dataSource: DataSourceType, dataSet: DataSetType) => {
+      const optionContent = {
+        type: dataSource.getType(),
+        name: dataSource.getName(),
+        ds: dataSource,
+      };
+      if (isIndexPatterns(dataSet)) {
+        const ip = dataSet as IndexPatternOption;
+        return {
+          ...optionContent,
+          label: ip.title,
+          value: ip.id,
+        };
+      }
+      return {
+        ...optionContent,
+        label: dataSource.getName(),
+        value: dataSource.getName(),
+      };
+    };
+
+    // Convert fetched datasets into a structured format suitable for selector rendering.
+    const getSourceList = (allDataSets: ISourceDataSet[]) => {
+      const finalList = [] as DataSourceGroup[];
+      allDataSets.forEach((curDataSet) => {
+        const typeKey = curDataSet.ds.getType() as DataSourceTypeKey;
+        const groupName = DATASOURCE_TYPE_DISPLAY_NAME_MAP[typeKey] || 'Default Group';
+
+        const existingGroup = finalList.find((item) => item.label === groupName);
+        const mappedOptions = curDataSet.data_sets?.map((dataSet) =>
+          getSourceOptions(curDataSet.ds, dataSet)
+        );
+
+        // check if add new datasource group or add to existing one
+        if (existingGroup) {
+          const existingOptionIds = new Set(existingGroup.options.map((opt) => opt.label));
+          const nonDuplicateOptions = mappedOptions.filter(
+            (opt) => !existingOptionIds.has(opt.label)
+          );
+          existingGroup.options.push(...nonDuplicateOptions);
+        } else {
+          finalList.push({
+            label: groupName,
+            options: mappedOptions,
+          });
+        }
+      });
+      return finalList;
+    };
+
+    Promise.all(fetchDataSets())
+      .then((results) => {
+        setDataSourceOptionList(getSourceList(results));
+      })
+      .catch((e) => onFetchDataSetError(e));
+  }, [dataSources, setDataSourceOptionList, onFetchDataSetError]);
+
+  const handleSourceChange = (selectedOptions: any) => onDataSourceSelect(selectedOptions);
```

Comment:
could you explain a bit more?

Reply:
discussed offline, you want to memorize `handleSourceChange`. But i still didn't understand how it's memory leak? 

The function is passed to `onChange` to combo_box and then combo_box_input. During rerender it is created again but the old function gets garbage collected. i looked into combo box code and didn't see anything relying on `props.onChange` being the same reference. Am i missing something?

Reply:
Not sure what i was thinking; probably was half asleep. This is not adding an event-listener.

Reply:
discussed offline, you wanted us to add `useCallback`. 

according to docs https://react.dev/reference/react/useCallback#should-you-add-usecallback-everywhere

> Caching a function with useCallback  is only valuable in a few cases:
> - You pass it as a prop to a component wrapped in [memo.](https://react.dev/reference/react/memo) You want to skip re-rendering if the value hasn’t changed. Memoization lets your component re-render only if dependencies changed.
> - The function you’re passing is later used as a dependency of some Hook. For example, another function wrapped in useCallback depends on it, or you depend on this function from [useEffect.](https://react.dev/reference/react/useEffect)

So my point is:
1. it is not a memory leak, because of gc
2. `useCallback` does not help in this case, because based on what i searched (i could be wrong due to how i search, but please point out with reference links),
    1. `handleSourceChange` is passed into `EuiComboBox`, which is not wrapped in memo, so the first point does not apply
    2. `EuiComboBox` does not use this as a dependency of a hook, so the second point does not apply
3. i believe internally react will still create the callback, and checks if dependency changed
    1. if dependency changed, then the new callback is returned
    2. if dependency did not change, then the memroized callback is returned and new callback is discarded
    3. i don't have reference on this unfortunately, but memorization definitely comes with a cost